newIDE/app/src/UI/SemiControlledAutoComplete.js

```diff
@@ -137,219 +69,192 @@ type State = {|
   inputValue: string | null,
 |};
 
-/**
- * An autocomplete field, showing options as the user type (or when the user presses down button).
- * Options can be chosen with keyboard or mouse.
- *
- * Supports divider between items and special items with `onClick` prop that when present is called
- * when the item is selected (and value is not changed).
- */
 export default class SemiControlledAutoComplete extends React.Component<
   Props,
   State
 > {
   _input = React.createRef<HTMLInputElement>();
+
   state = {
     inputValue: null,
   };
 
-  /**
-   * Allow to override the value being written in the autocomplete, if any.
-   * Usually you don't want to do that as the point of having a "SemiControlled" auto-complete
-   * is that the value being written by the user is the source of truth.
-   *
-   * In some cases, you want to force a new value when the auto-complete is focused (for example,
-   * if you opened a native dialog selector that is going to make the auto-complete focused/blurred/
-   * focused again, in a manner that depends on the OS).
-   * Call this function to do so.
-   *
-   * @param newValue The new value to set in the auto complete. If the field is not being focused,
-   * nothing will be forced (the value props will be used when the field is focused).
-   */
-  forceInputValueTo(newValue: string) {
+  focus() {
+    if (this._input.current) this._input.current.focus();
+  }
+
+  forceInputValueTo(newValue: string): void {
     if (this.state.inputValue !== null) {
       this.setState({
         inputValue: newValue,
       });
     }
   }
 
-  focus() {
-    if (this._input.current) this._input.current.focus();
-  }
+  isOptionDisabled = (option: Option): boolean => {
+    // Writing shorthand, ie, return option.type === 'separator' fails flow test.
+    // All the cases of type Option have to be explicitly handled.
+    if (option.type === 'separator') return true;
+    else return false;
+  };
 
   render() {
-    const { props, state } = this;
-    const currentInputValue =
-      state.inputValue !== null ? state.inputValue : props.value;
+    const getCurrentInputValue = (): string =>
+      this.state.inputValue !== null ? this.state.inputValue : this.props.value;
+
+    const filterFunction = (options: DataSource, state: Object): DataSource => {
+      var separatorCount = 0;
+      const lowercaseInputValue = getCurrentInputValue().toLowerCase();
+      var optionList = options.filter(option => {
+        if (option.type === 'separator') {
+          separatorCount++;
+          return true;
+        }
+        if (!option.text) return true;
+        return option.text.toLowerCase().indexOf(lowercaseInputValue) !== -1;
+      });
 
-    const helperText = props.helperMarkdownText ? (
-      <MarkdownText source={props.helperMarkdownText} />
+      // Empty the list if there are only dividers inside it
+      if (optionList.length === separatorCount) optionList = [];
```

Comment:
Could surely be rewritten as:
```
// Show an empty list if there are only separators.
if (!filteredOptionsList.filter(option => option.type !== 'separator').length) return [];
```

Which avoids the need to do the `separatorCount` and clearly separate the two concerns:
* first you filter the list according to the input value
* then you check if there are results, if not, you bail out.

Reply:
So this should be the first condition in that function, right? And after checking this, the list should get filtered?

`if(list contains only separator) return []` (bailing out)
`let count = 0`
`filter the list`
`delete separators at the start and end of array`

Is this correct?

Reply:
Alright, so I did what I said above. Let me know if its the correct way or I misunderstood you.
Also I had to write this,
`if (!options.filter(option => option.type !== 'separator' && option.text).length) return [];`

(it took me more than 15 mins to figure out how to make flow stop complaining :p I had to add a redundant explicit check)

Reply:
> if(list contains only separator) return [] (bailing out)
> filter the list
> delete separators at the start and end of array

You want to filter the list first no?
First you filter the list (that's the core filtering logic), and then you apply some beautifying stuff (i.e: you return an empty list if you have only separators and then you remove the separator that are at the end/beginning)